components/org.wso2.carbon.identity.cors.valve/src/main/java/org/wso2/carbon/identity/cors/valve/CORSValve.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * NOTE: The code/logic in this class is copied from https://bitbucket.org/thetransactioncompany/cors-filter.
+ * All credits goes to the original authors of the project https://bitbucket.org/thetransactioncompany/cors-filter.
+ */
+
+package org.wso2.carbon.identity.cors.valve;
+
+import org.apache.catalina.connector.Request;
+import org.apache.catalina.connector.Response;
+import org.apache.catalina.valves.ValveBase;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.cors.mgt.core.exception.CORSManagementServiceException;
+import org.wso2.carbon.identity.cors.valve.exception.CORSValveException;
+import org.wso2.carbon.identity.cors.valve.internal.handler.CORSRequestHandler;
+import org.wso2.carbon.identity.cors.valve.internal.util.CORSUtils;
+import org.wso2.carbon.identity.cors.valve.internal.wrapper.CORSResponseWrapper;
+import org.wso2.carbon.identity.cors.valve.model.CORSRequestType;
+
+import java.io.IOException;
+
+import javax.servlet.ServletException;
+
+/**
+ * Valve class for CORS.
+ */
+public class CORSValve extends ValveBase {
+
+    private static final Log log = LogFactory.getLog(CORSValve.class);
+
+    private final CORSRequestHandler corsRequestHandler;
+
+    /**
+     * Default constructor for the CORSValve.
+     */
+    public CORSValve() {
+
+        corsRequestHandler = new CORSRequestHandler();
+    }
+
+    @Override
+    public void invoke(Request request, Response response) throws IOException, ServletException {
+
+        // Determine the type of the request.
+        CORSRequestType corsRequestType = CORSUtils.getRequestType(request);
+
+        try {
+            if (corsRequestType == CORSRequestType.ACTUAL) {
+                // ACTUAL request.
+                corsRequestHandler.handleActualRequest(request, response);
+
+                // Preserve CORS response headers on reset()
+                // TODO
+                CORSResponseWrapper responseWrapper = new CORSResponseWrapper(response);
+
+                getNext().invoke(request, response);
+            } else if (corsRequestType == CORSRequestType.PREFLIGHT) {
+                // PREFLIGHT request.
+                corsRequestHandler.handlePreflightRequest(request, response);
```

Comment:
Don't we need `getNext().invoke(request, response);` after this

Reply:
Preflight request doesn't need to be propagated to the other layers.